### ISSUE TITLE
Feature request: rebuild documentation for old crates

### ISSUE THREAD

**Issue by tspiteri** [rust-lang/docs.rs#464]
When crates are not updated in a long time, their style starts to drift from the latest style (or the latest style drifts for them) and consequently they do not gain the latest rustdoc features. I think it would be nice if crates which were last updated one or two years ago are rerun and the docs updated with the latest rustdoc.

This could be low priority so that it does not slow down documentation of new crates.

If a compilation fails for whatever reason, of course the old version should be retained.

**Comment by jyn514** [comment]
Hmm I don't mind this idea but it seems expensive to rerun the build on all versions that have ever been published ... This would just affect style sheets and such, right? What should we do if building with a newer version of `rustdoc` leads to broken links?

**Comment by jyn514** [comment]
Related: https://github.com/rust-lang/docs.rs/issues/301

**Comment by jyn514** [comment]
https://github.com/rust-lang/docs.rs/issues/295 also mentioned allowing crate owners to re-queue builds for their own crates, e.g. if the build succeeds but the JavaScript throws a runtime error because of a bad nightly.

**Comment by jyn514** [comment]
The reason I originally thought this wasn't feasible is because rustdoc puts out releases very frequently (either daily or every 6 weeks, depending how you count it). But nothing says we have to rebuild for every release - if we rebuilt, say every edition, that would still fix most of the rustdoc bugs while being feasible to finish all crates.

Another thing we could do is only rebuild the crates that are used the most: say the latest version and most recent release of every crate. That would bring the number of builds down from ~220k to between 40-80k builds.

This shouldn't be implemented until after #1004 to avoid very large re-upload costs (@pietroalbini estimates the current cost of reuploading every crate at $2k even if there are no changes in the documentation).

**Comment by Kixiron** [comment]
I think with good culling parameters we could significantly reduce the number of crates needed to be built & uploaded
- Only the most recent release, pretty much required for feasibility
- Download/Visit thresholds. If we set cutoffs on crates that haven't been touched in years, no one really loses anything
- Don't rebuild the past few months of releases. The most recent builds will look pretty much the same, bumping a few nightlies shouldn't affect anything. It'll also let us avoid the "edition rush" of crates that are preparing for the new edition

I really want to propose some sort of documentation hashing/checksuming to attempt to find out if the docs currently uploaded differ from the ones we build and then don't delete & reupload if there's no change, but I doubt that's possible due to the unstable nature of rustdoc's output

**Comment by jyn514** [comment]
(changing the title because I got tired of never being able to find it when searching 😆 )

**Comment by shepmaster** [comment]
> * Download/Visit thresholds. If we set cutoffs on crates that haven't been touched in years, no one really loses anything

This was my thought. If a crate gets a lot of visits, then it's worth rebuilding the docs. In my particular instance, the futures crate docs appears to suffer from a bug where the `<'a>` was left off a trait method, which makes it look like maybe the lifetime comes from the trait instead.

You could bucket things up, as well. I don't know the actual limits in place, but something like:

Get the number of visits to a crate's docs between _stable_ Rust release dates.

For the top X% of crates, regenerate them if they haven't been naturally generated in the last Y stable releases.

|X|Y|
|-|-|
|10|1| 
|50|3| 
|90|6|

**Comment by jyn514** [comment]
Stable releases seems like an arbitrary metric, since crates are always built with nightly. We could use 6 weeks to sync with the release schedule, but I'm a little worried people will be confused if there's a new bug introduced on nightly or something.

In any case, I would prefer not to make major changes here until https://github.com/rust-lang/docs.rs/issues/1004 is fixed.

**Comment by ShadowJonathan** [comment]
One other thing that i'd like to suggest is rebuild requests, to be able to manually request somewhere to rebuild a (very) out of date documentation of a crate.

I just encountered this with [`anymap`](https://docs.rs/anymap/0.12.1/anymap/), and while i think a rebuild of old documentation on every major release, or after a amount of "rotting" time (lets say, longer than a year or two years), and then automatically rebuilding them (most popular crates first), while i think that would be a good approach, the best heuristic would be to be able to have users re-request a build. (and have the button visible and active only when a build is X amount of time old)

In a good-faith case, someone clicks this when they see the documentation style outdated, and they think it'd be "good" to have it be up-to-date.

In a bad faith case, someone goes around queuing this for every crate. I personally see this as alright, but it depends on how much the build system can handle, and if it's possible to queue this on a very low priority.

**Comment by jyn514** [comment]
@ShadowJonathan see also https://github.com/rust-lang/docs.rs/issues/301

**Comment by ShadowJonathan** [comment]
@GuillaumeGomez @syphar thanks a lot!
